File: python/Gaffer/Path.py
##########################################################################
#  
#  Copyright (c) 2011, John Haddon. All rights reserved.
#  Copyright (c) 2011-2013, Image Engine Design Inc. All rights reserved.
#  
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#  
#      * Redistributions of source code must retain the above
#        copyright notice, this list of conditions and the following
#        disclaimer.
#  
#      * Redistributions in binary form must reproduce the above
#        copyright notice, this list of conditions and the following
#        disclaimer in the documentation and/or other materials provided with
#        the distribution.
#  
#      * Neither the name of John Haddon nor the names of
#        any other contributors to this software may be used to endorse or
#        promote products derived from this software without specific prior
#        written permission.
#  
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#  
##########################################################################

import warnings

import Gaffer

class Path( object ) :

	## Paths can be constructed either from a "/slash/separated/string"
	# or a [ "list", "of", "strings" ] relative to a root.
	## \todo Reimplement the class paths in C++, with three distinct
	# constructors :
	#
	# Path( FilterPtr filter = 0 ) // empty path
	# Path( const std::string &path, FilterPtr filter = 0 )
	# Path( const std::vector<InternedString> &path, const std::string &root="/", FilterPtr filter = 0 )
	def __init__( self, path=None, root="/", filter=None ) :
	
		assert( isinstance( root, basestring ) )
	
		self.__items = []
		self.__root = ""
		
		if isinstance( path, basestring ) :
			
			self.setFromString( path )
			
		elif path is not None :
		
			self.__root = root
			for p in path :
					
				self.__checkElement( p )		
				self.__items.append( p )
				
		self.__filter = None
		self.setFilter( filter )
	
	## Returns the root of the path - this will be "/" for absolute
	# paths and "" for relative paths.
	def root( self ) :
	
		return self.__root
	
	## Returns true if this path is empty.
	def isEmpty( self ) :
	
		return not len( self ) and not self.__root
									
	## Returns true if this path is valid - ie references something
	# which actually exists.
	def isValid( self ) :
		
		return not self.isEmpty()
	
	## If the path is valid, returns a dictionary of information about what
	# the path points to. If the path is not valid, returns None. The contents
	# of the dictionary depends on the type of Path. Subclasses should call
	# their base class' implementation first, and then add additional information
	# to the dictionary returned.
	def info( self ) :
	
		if not self.isValid() :
			return None
			
		result = {}
		result["name"] = self[-1] if len( self ) else ""
		result["fullName"] = str( self )
		
		return result
	
	## Returns true if this path can never have child Paths.
	def isLeaf( self ) :
	
		raise NotImplementedError
	
	## Returns the parent of this path, or None if the path
	# has no parent (is the root).
	def parent( self ) :
	
		if not self.__items :
			return None
			
		parent = self.copy()
		del parent[-1]
		
		return parent
	
	## Returns a list of Path instances representing all
	# the children of this path. Note that an empty list may
	# be returned even if isLeaf() is False.
	def children( self ) :
	
		c = self._children()
			
		if self.__filter is not None :
			c = self.__filter.filter( c )
			
		return c	
	
	## The subclass specific part of children(). This must be implemented
	# by subclasses to return a list of children - filtering will be applied
	# in the children() method so can be ignored by the derived classes.	
	def _children( self ) :
	
		raise NotImplementedError
	
	def setFilter( self, filter ) :
	
		if filter is self.__filter :
			return
			
		self.__filter = filter
		if self.__filter is not None :
			self.__filterChangedConnection = self.__filter.changedSignal().connect( Gaffer.WeakMethod( self.__filterChanged ) )
		else :
			self.__filterChangedConnection = None
			
		self.__emitChangedSignal()
		
	def getFilter( self ) :
	
		return self.__filter
	
	## \deprecated
	# Use setFilter() instead. If you wish to use more than
	# one filter then use a CompoundPathFilter.
	def addFilter( self, pathFilter ) :
	
		warnings.warn( "Path.addFilter() is deprecated, use Path.setFilter() instead. If you wish to use more than one filter then use a CompoundPathFilter.", DeprecationWarning, 2 )		
		
		if not isinstance( self.__filter, Gaffer.CompoundPathFilter ) :
			self.__filter = Gaffer.CompoundPathFilter()	
			
		self.__filter.addFilter( pathFilter )
		
	## \deprecated
	# Use setFilter() instead. If you wish to use more than
	# one filter then use a CompoundPathFilter.
	def removeFilter( self, pathFilter ) :
	
		warnings.warn( "Path.removeFilter() is deprecated, use Path.setFilter() instead. If you wish to use more than one filter then use a CompoundPathFilter.", DeprecationWarning, 2 )		

		self.__filter.removeFilter( pathFilter )

	def pathChangedSignal( self ) :
	
		if not hasattr( self, "_Path__pathChangedSignal" ) :
			self.__pathChangedSignal = Gaffer.Signal1()
			
		return self.__pathChangedSignal

	## Sets the path root and items from the other
	# path, leaving the current filter intact.
	def setFromPath( self, path ) :
	
		if path.__items == self.__items and path.__root == self.__root :
			return
			
		self.__items = path.__items[:]
		self.__root = path.__root
		self.__emitChangedSignal()
		
	## Sets the path root and items from a "/"
	# separated string.
	def setFromString( self, path ) :
		
		newItems = [ x for x in path.split( "/" ) if x ]
		newRoot = "/" if path and path[0]=='/' else ""
				
		if newItems != self.__items or newRoot != self.__root :
			self.__items = newItems
			self.__root = newRoot
			self.__emitChangedSignal()
			
		return self
	
	def copy( self ) :
	
		c = self.__class__( self.__items, root=self.__root )
		c.setFilter( self.__filter )
		
		return c
		
	def append( self, element ) :
	
		self.__checkElement( element )
		self.__items.append( element )
		self.__emitChangedSignal()
		
		return self
	
	def truncateUntilValid( self ) :
	
		changed = False
		while len( self.__items ) and not self.isValid() :
			del self.__items[-1]
			changed = True
			
		if changed :
			self.__emitChangedSignal()
		
		return self
		
	def __len__( self ) :
	
		return len( self.__items )

	def __str__( self ) :
				
		return self.__root + "/".join( self.__items )
			
	def __setitem__( self, index, name ) :
	
		if isinstance( index, slice ) :
			for n in name :
				self.__checkElement( n )
		else :
			self.__checkElement( name )
		
		prev = self.__items[index]
				
		self.__items.__setitem__( index, name )
		
		if prev!=name :
			self.__emitChangedSignal()
		
	def __getitem__( self, index ) :
	
		return self.__items.__getitem__( index )

	def __delitem__( self, index ) :
	
		self.__items.__delitem__( index )
		self.__emitChangedSignal()
		
	def __eq__( self, other ) :
	
		if not isinstance( other, Path ) :
			return False
			
		return self.__items == other.__items and self.__root == other.__root
		
	def __ne__( self, other ) :
	
		if not isinstance( other, Path ) :
			return True
			
		return self.__items != other.__items	
			
	def __checkElement( self, element ) :
	
		if not isinstance( element, basestring ) :
			raise ValueError( "Path elements must be strings." )
				
		if "/" in element :
			raise ValueError( "Path element contains \"/\"." )
		
		if element=="" :
			raise ValueError( "Path element is empty." )	

	def __emitChangedSignal( self ) :
		
		if hasattr( self, "_Path__pathChangedSignal" ) :
			self.__pathChangedSignal( self )
			
	def __filterChanged( self, filter ) :
	
		assert( filter is self.__filter )
		
		self.__emitChangedSignal()
		
	def __repr__( self ) :
	
		return "%s( '%s' )" % ( self.__class__.__name__, str( self ) )

File: src/Gaffer/Path.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2011, John Haddon. All rights reserved.
//  Copyright (c) 2011-2013, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//      * Redistributions of source code must retain the above
//        copyright notice, this list of conditions and the following
//        disclaimer.
//
//      * Redistributions in binary form must reproduce the above
//        copyright notice, this list of conditions and the following
//        disclaimer in the documentation and/or other materials provided with
//        the distribution.
//
//      * Neither the name of John Haddon nor the names of
//        any other contributors to this software may be used to endorse or
//        promote products derived from this software without specific prior
//        written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include "Gaffer/Path.h"

#include "Gaffer/PathFilter.h"

#include "IECore/SimpleTypedData.h"
#include "IECore/StringAlgo.h"

#include "boost/bind.hpp"

using namespace std;
using namespace IECore;
using namespace Gaffer;

IE_CORE_DEFINERUNTIMETYPED( Path );

static InternedString g_namePropertyName( "name" );
static InternedString g_fullNamePropertyName( "fullName" );

Path::Path( PathFilterPtr filter )
	:	m_pathChangedSignal( nullptr )
{
	setFilter( filter );
}

Path::Path( const std::string &path, PathFilterPtr filter )
	:	m_pathChangedSignal( nullptr )
{
	setFromString( path );
	setFilter( filter );
}

Path::Path( const Names &names, const IECore::InternedString &root, PathFilterPtr filter )
	:	m_root( root ), m_names( names ), m_pathChangedSignal( nullptr )
{
	for( Names::const_iterator it = m_names.begin(), eIt = m_names.end(); it != eIt; ++it )
	{
		checkName( *it );
	}
	setFilter( filter );
}

Path::~Path()
{
	if( havePathChangedSignal() && m_filter )
	{
		// In an ideal world, we'd derive from Signals::Trackable, and wouldn't
		// need to do this manual connection management. But we construct a lot of Path
		// instances and trackable has significant overhead so we must avoid it. We must
		// disconnect somehow though, otherwise when the filter changes, filterChanged()
		// will be called on a dead Path instance.
		/// \todo Review this decision - it was made when we were using `boost::signals::trackable`
		/// and may not longer be valid.
		m_filter->changedSignal().disconnect( boost::bind( &Path::filterChanged, this ) );
	}
	delete m_pathChangedSignal;
}

const IECore::InternedString &Path::root() const
{
	return m_root;
}

bool Path::isEmpty() const
{
	return m_names.empty() && m_root.string().empty();
}

bool Path::isValid( const IECore::Canceller *canceller ) const
{
	return !isEmpty();
}

bool Path::isLeaf( const IECore::Canceller *canceller ) const
{
	return false;
}

void Path::propertyNames( std::vector<IECore::InternedString> &names, const IECore::Canceller *canceller ) const
{
	names.push_back( g_namePropertyName );
	names.push_back( g_fullNamePropertyName );
}

IECore::ConstRunTimeTypedPtr Path::property( const IECore::InternedString &name, const IECore::Canceller *canceller ) const
{
	if( name == g_namePropertyName )
	{
		if( m_names.size() )
		{
			return new StringData( m_names.back().string() );
		}
		else
		{
			return new StringData( "" );
		}
	}
	else if( name == g_fullNamePropertyName )
	{
		return new StringData( this->string() );
	}
	return nullptr;
}

PathPtr Path::parent() const
{
	if( m_names.empty() )
	{
		return nullptr;
	}

	PathPtr result = copy();
	result->m_names.pop_back();

	return result;
}

size_t Path::children( std::vector<PathPtr> &children, const IECore::Canceller *canceller ) const
{
	doChildren( children, canceller );
	if( m_filter )
	{
		m_filter->filter( children, canceller );
	}
	return children.size();
}

void Path::setFilter( PathFilterPtr filter )
{
	if( filter == m_filter )
	{
		return;
	}

	// We may need to connect to Filter::changedSignal() so we can
	// emit pathChangedSignal() appropriately. But it's not worth making
	// the connection unless m_pathChangedSignal exists.
	if( havePathChangedSignal() )
	{
		if( m_filter )
		{
			m_filter->changedSignal().disconnect( boost::bind( &Path::filterChanged, this ) );
		}
		if( filter )
		{
			filter->changedSignal().connect( boost::bind( &Path::filterChanged, this ) );
		}
	}

	m_filter = filter;
	emitPathChanged();
}

PathFilter *Path::getFilter()
{
	return m_filter.get();
}

const PathFilter *Path::getFilter() const
{
	return m_filter.get();
}

Path::PathChangedSignal &Path::pathChangedSignal()
{
	if( !m_pathChangedSignal )
	{
		m_pathChangedSignal = new PathChangedSignal;
		pathChangedSignalCreated();
	}
	return *m_pathChangedSignal;
}

void Path::setFromPath( const Path *path )
{
	if( path->m_names == m_names && path->m_root == m_root )
	{
		return;
	}

	m_root = path->m_root;
	m_names = path->m_names;

	emitPathChanged();
}

void Path::setFromString( const std::string &string )
{
	Names newNames;
	StringAlgo::tokenize<InternedString>( string, '/', back_inserter( newNames ) );

	InternedString newRoot;
	if( string.size() && string[0] == '/' )
	{
		newRoot = "/";
	}

	if( newRoot == m_root && newNames == m_names )
	{
		return;
	}

	m_names = newNames;
	m_root = newRoot;

	emitPathChanged();
}

PathPtr Path::copy() const
{
	return new Path( m_names, m_root, m_filter );
}

void Path::append( const IECore::InternedString &name )
{
	checkName( name );
	m_names.push_back( name );
	emitPathChanged();
}

void Path::truncateUntilValid()
{
	bool changed = false;
	while( m_names.size() && !isValid() )
	{
		m_names.pop_back();
		changed = true;
	}

	if( changed )
	{
		emitPathChanged();
	}
}

const Path::Names &Path::names() const
{
	return m_names;
}

void Path::set( size_t index, const IECore::InternedString &name )
{
	if( index >= m_names.size() )
	{
		throw IECore::Exception( "Index out of range" );
	}

	if( name == m_names[index] )
	{
		return;
	}

	checkName( name );

	m_names[index] = name;
	emitPathChanged();
}

void Path::set( size_t begin, size_t end, const Names &names )
{
	if( begin > m_names.size() )
	{
		throw IECore::Exception( "Index out of range" );
	}

	if( end > m_names.size() )
	{
		throw IECore::Exception( "Index out of range" );
	}

	Names::difference_type sizeDifference = names.size() - (end - begin);

	if( sizeDifference == 0 )
	{
		if( equal( m_names.begin() + begin, m_names.begin() + end, names.begin() ) )
		{
			return;
		}
	}
	else if( sizeDifference > 0 )
	{
		m_names.resize( m_names.size() + sizeDifference );
		std::copy_backward( m_names.begin() + end, m_names.begin() + end + sizeDifference, m_names.end() );
	}
	else
	{
		std::copy( m_names.begin() + end, m_names.end(), m_names.begin() + end + sizeDifference );
		m_names.resize( m_names.size() + sizeDifference );
	}

	std::copy( names.begin(), names.end(), m_names.begin() + begin );
	emitPathChanged();
}

void Path::remove( size_t index )
{
	if( index >= m_names.size() )
	{
		throw IECore::Exception( "Index out of range" );
	}

	m_names.erase( m_names.begin() + index );
	emitPathChanged();
}

void Path::remove( size_t begin, size_t end )
{
	if( begin >= m_names.size() )
	{
		throw IECore::Exception( "Index out of range" );
	}

	if( end > m_names.size() )
	{
		throw IECore::Exception( "Index out of range" );
	}

	m_names.erase( m_names.begin() + begin, m_names.begin() + end );
	emitPathChanged();
}

std::string Path::string() const
{
	std::string result = m_root.string();
	for( size_t i = 0, s = m_names.size(); i < s; ++i )
	{
		if( i != 0 )
		{
			result += "/";
		}
		result += m_names[i].string();
	}
	return result;
}

bool Path::operator == ( const Path &other ) const
{
	return
		typeId() == other.typeId() &&
		m_root == other.m_root &&
		m_names == other.m_names;
}

bool Path::operator != ( const Path &other ) const
{
	return !(*this == other );
}

const Plug *Path::cancellationSubject() const
{
	return nullptr;
}

void Path::doChildren( std::vector<PathPtr> &children, const IECore::Canceller *canceller ) const
{
}

void Path::emitPathChanged()
{
	if( !m_pathChangedSignal )
	{
		return;
	}
	(*m_pathChangedSignal)( this );
}

void Path::pathChangedSignalCreated()
{
	if( m_filter )
	{
		m_filter->changedSignal().connect( boost::bind( &Path::filterChanged, this ) );
	}
}

bool Path::havePathChangedSignal() const
{
	return m_pathChangedSignal;
}

void Path::filterChanged()
{
	emitPathChanged();
}

void Path::checkName( const IECore::InternedString &name ) const
{
	if( name.string().find( '/' ) != string::npos )
	{
		throw IECore::Exception( "Path name contains '/'." );
	}

	if( name.string().empty() )
	{
		throw IECore::Exception( "Path name is empty." );
	}
}
